core/service_test.go

```diff
@@ -0,0 +1,363 @@
+			[]string{
+				"MsgUpdateClient(dstClient)",
+				"MsgRecvPacket(1)",
+			},
+		},
+		"timeout": { // timeout. Relay back to src channel as MsgTimeout with UpdateClient.
+			"ORDERED",
+			1,
+			[]*core.PacketInfo{
+				newPacketInfo(1, 90),
+			},
+			[]string{ "MsgUpdateClient(srcClient)", "MsgTimeout(1)" },
+			[]string{ },
+		},
+		"timeout at latest block but not at finalized block(at lower border)": { // waiting relay in finalized block
+			"ORDERED",
+			1,
+			[]*core.PacketInfo{
+				newPacketInfo(1, 91),
+			},
+			[]string{ },
+			[]string{ },
+		},
+		"timeout at latest block but not at finalized block(at heigher border)": { // waiting relay in finalized block
+			"ORDERED",
+			1,
+			[]*core.PacketInfo{
+				newPacketInfo(1, 100),
+			},
+			[]string{ },
+			[]string{ },
+		},
+		"multiple timeouts packets in ordered channel": { // In ordered channel, later packets from timeouted packets are not relayerd
+			"ORDERED",
+			1,
+			[]*core.PacketInfo{
+				newPacketInfo(1, 9),
+				newPacketInfo(2, 9999),
+				newPacketInfo(3, 9),
+			},
+			[]string{
+				"MsgUpdateClient(srcClient)",
+				"MsgTimeout(1)",
+			},
+			[]string{
+			},
+		},
+		"relay preceding packets before timeouted one": { // In ordered channel, only preceding packets before timeout packets are relayed.
+			"ORDERED",
+			1,
+			[]*core.PacketInfo{
+				newPacketInfo(1, 9999),
+				newPacketInfo(2, 9999),
+				newPacketInfo(3, 9),
+			},
+			[]string{ },
+			[]string{
+				"MsgUpdateClient(dstClient)",
+				"MsgRecvPacket(1)",
+				"MsgRecvPacket(2)",
+			},
+		},
+		"multiple timeout packets in unordered channel": { // In unordered channel, all timeout packets are backed and others are relayed.
+			"UNORDERED",
+			1,
+			[]*core.PacketInfo{
+				newPacketInfo(1, 9999),
+				newPacketInfo(2, 9),
+				newPacketInfo(3, 9999),
+				newPacketInfo(4, 9),
+			},
+			[]string{
+				"MsgUpdateClient(srcClient)",
+				"MsgTimeout(2)",
+				"MsgTimeout(4)",
+			},
+			[]string{
+				"MsgUpdateClient(dstClient)",
+				"MsgRecvPacket(1)",
+				"MsgRecvPacket(3)",
+			},
+		},
+	}
+	for n, c := range cases {
+		if n[0] == '_' { continue }
+		t.Run(n, func (t2 *testing.T) { testServe(t2, c) })
+	}
+}
+
+func testServe(t *testing.T, tc testCase) {
+	log.InitLoggerWithWriter("debug", "text", os.Stdout, false)
+	telemetry.InitializeMetrics()
+
+	srcLatestHeader := mocktypes.Header{
+		Height: clienttypes.NewHeight(1, 100),
+		Timestamp: uint64(10100),
+	}
+	dstLatestHeader := mocktypes.Header{
+		Height: clienttypes.NewHeight(1, 100),
+		Timestamp: uint64(10100),
+	}
+
+	ctrl := gomock.NewController(t)
+
+	var unreceivedPackets []uint64
+	for _, p := range tc.UnfinalizedRelayPackets {
+		unreceivedPackets = append(unreceivedPackets, p.Sequence)
+	}
+	src := NewMockProvableChain(ctrl, "src", tc.Order, srcLatestHeader, tc.UnfinalizedRelayPackets, []uint64{})
+	dst := NewMockProvableChain(ctrl, "dst", tc.Order, dstLatestHeader, []*core.PacketInfo{}, unreceivedPackets)
```

Comment:
This code always sets no packet backlog for the destination chain, but we need some test cases where the both chains have packet backlog.